[김은혜] Step1 지하철역 인수 테스트 작성 PR 요청

eunveloper opened the pull request:
안녕하세요! 인수테스트 파트를 함께하게 되어서 정말 영광입니다 :)

병렬적으로 과제를 진행하게 되었는데, 중간에 강의를 듣고 수정을 하고싶은 부분이 생기면
수정해서 또 PR을 날릴테니, 너무 귀찮아하지 말아주세요 😋

가감없는 피드백 잘 부탁드립니다~ 많이 배워가겠습니다!

eunveloper committed:
decs: README.md 작성

- 1단계 (지하철역 인수 테스트 작성) 요구사항 작성
- 인수테스트 조건명세 (등록, 조회, 삭제)

eunveloper committed:
feat: 지하철역 인수 테스트 작성

- 지하철역 조회 인수 테스트 작성
- 지하철역 삭제 인수 테스트 작성 (예외 케이스 추가)

geunhwanlee commented on README.md:
```diff
@@ -1,54 +1,49 @@
+- Feature: 지하철역을 등록한다.
+- Scenario: 새로운 지하철역 등록
+- Given: 등록하고자 하는 지하철역이 미등록 상태이다.
+- And: 그외의 지하철역은 등록되어 있어도 무방하다.
```
```
And: 그외의 지하철역은 등록되어 있어도 무방하다.
```

제 생각에는 인수 조건과 무관한 항목인것 같아요! 🤔

은혜씨는 어떻게 생각하시는지 궁금합니다. 🙏

eunveloper replied:
아 저는 예외케이스의 부분을 고려하여서 작성 하였는데,
사실 인수테스트의 조건과는 무관하긴 한거같아요! 

차라리 저런 예외케이스의 부분은 추가로 예외케이스를 테스트하는
추가 인수 테스트를 작성하는게 좋을것 같네요! 🥹

geunhwanlee commented on README.md:
```diff
@@ -1,54 +1,49 @@
+- Given: 삭제하고자 하는 지하철역이 등록 상태이다.
+- And: 그외의 지하철역은 등록되어 있지 않아도 무방하다.
+- When: 사용자는 삭제하고자 하는 지하철역의 정보를 입력하고 삭제를 요청한다.
+- Then: 삭제가 정상적으로 됐다는 응답을받는다.
```
인수테스트 시나리오를 잘 작성해주셨네요. 👍🏻

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -101,6 +102,38 @@ void createStationWithDuplicateName() {
     @Test
     void getStations() {
+        // given: 2개의 지하철역이 등록되어 있다.
+        Map<String, String> param = new HashMap<>();
+
+        param.put("name", "강남역");
+        RestAssured.given().log().all()
+                .body(param)
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .when().post("/stations")
+                .then().log().all();
```
중복되는 코드가 반복되고있네요. 🤔

메서드로 추출하면 중복을 제거할수 있고
시나리오의 흐름을 가독성있게 표현할수 있겠네요! 👍🏻

eunveloper replied:
가장 기본적인 중복코드 제거를 놓치다니!
기본이 중요하다는 것을 다시한번 배워가요 🤣

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -101,6 +102,38 @@ void createStationWithDuplicateName() {
+                .when().post("/stations")
+                .then().log().all();
+
+        param.put("name", "홍대역");
+        RestAssured.given().log().all()
+                .body(param)
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .when().post("/stations")
+                .then().log().all();
```
지하철역 생성후 정상적으로 생성됐는지 확인하기 위해
statusCode를 체크해보는건 어떨까요? 🤔

eunveloper replied:
최종 케이스만 statusCode를 검증했는데, 
중간에 when의 경우에도 잘 생성됐는지 확인하는게 좋을거 같아요!

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -101,6 +102,38 @@ void createStationWithDuplicateName() {
+        assertAll(
+                () -> assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value()),
+                () -> assertThat(stationNames.size()).isEqualTo(2)
+        );
```
조회된 지하철역의 이름도 검증해보는건 어떨까요?

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -111,5 +144,57 @@ void getStations() {
+                .when().post("/stations")
+                .then().log().all().extract();
+
+        Long id = created.response().body().jsonPath().getLong("id");
```
지하철역을 생성하면 응답 헤더의 location에 생성된 지하철역에 접근할수 있는 uri을 응답해주고있어요.
location으로 삭제 요청을 보내면 생성 api의 body와 무관하게 처리할수 있을것 같은데 어떻게 생각하시나요? 🤔

eunveloper replied:
앗, 응답데이터 분석이 좀 부족했던거 같네요 🥹

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -111,5 +144,57 @@ void getStations() {
+
+        Long id = created.response().body().jsonPath().getLong("id");
+
+        // when: 생성한 지하철을 삭제한다.
+        ExtractableResponse<Response> response =
+                RestAssured.given().log().all()
+                        .contentType(MediaType.APPLICATION_JSON_VALUE)
+                        .when().delete("/stations/" + id)
+                        .then()
+                        .extract();
```
삭제도 마찬가지로 statusCode를 검증해보면 좋을것같아요!

geunhwanlee reviewed (changes requested):
안녕하세요!
인수테스트 미션을 함께 하게될 이근환입니다. 🙌

1단계 인수테스트 작성 아주 잘해주셨어요!

몇가지 아쉬운 점이 있어서 코멘트 드렸습니다.

저는 개인적으로 시나리오형태의 인수테스트를 작성할때는
시나리오의 흐름이 잘 보이도록 작성하는게 중요하다고 생각해요.

예를 들면 지하철역 목록을 조회하는 시나리오에서
```
// given
지하철역 등록되어있음("강남역");
지하철역 등록되어있음("홍대역");

// when
조회결과 = 지하철역 조회();

// then
지하철역 조회성공(조회결과, "강남역", "홍대역");
```

이런식으로 흐름이 보이도록 메서드로 추출 했었어요! 😀
또한 중복코드도 제거가 됩니다. 👍🏻
정답은 없으니 참고정도만 부탁드립니다!

한번 고민해보시고 재 리뷰요청 주세요. 🙏

eunveloper committed:
refactor: 피드백 반영

- 불필요 인수조건 제거
- 중복되는 코드 추출 및 공통화
- API 호출후 StatusCode 검증
- 테스트 케이스 추가
- location 활용

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -74,20 +67,8 @@ void createStationWithDuplicateName() {
-                        .when().post("/stations")
-                        .then().log().all()
-                        .extract();
+        ExtractableResponse<Response> response = insertStation(params).extract();
```
테스트가 실패하네요. 😭
given에서 강남역을 생성해 둬야 할 것 같아요!

geunhwanlee commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -111,5 +113,67 @@ void getStations() {
     @Test
     void deleteStation() {
+        // given: 제거할 지하철 역을 생성한다.
+        Map<String, String> param = new HashMap<>();
+
+        param.put("name", "강남역");
+        ExtractableResponse<Response> created = insertStation(param).extract();
+        assertThat(created.statusCode()).isEqualTo(HttpStatus.CREATED.value());
```
```suggestion
        insertStationSuccess("강남역");
```
이렇게 메서드로 분리하는건 어떻게 생각하시나요? 🤔

eunveloper replied:
오호 이러한 부분도 메소드로 분리! 메소드 명이 명확해서 오히려 가독성도 좋아지고 코드도 깔끔한거 같아요!

geunhwanlee reviewed (approved):
1단계 인수 테스트 작성 잘 해주셨어요! 👍🏻

중복된 코드도 많이 제거가 됐네요 👏

다만 테스트 각각이 격리되어 있지 않아서 테스트끼리 서로 영향을 주고 있어요.
다음 단계에선 인수 테스트를 어떻게 격리할 수 있을지 고민해 보셨으면 좋겠습니다. 🙏

그럼 다음 단계도 화이팅 하세요 🔥
